File: idea-plugin/src/main/java/com/zzf/rikki/idea/agent/LiteAgentServer.java
package com.zzf.rikki.idea.agent;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.intellij.openapi.project.Project;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.HttpServer;
import com.zzf.rikki.idea.agent.compat.InMemoryPendingApprovalService;
import com.zzf.rikki.idea.agent.compat.LiteToolExecutor;

import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executors;

public class LiteAgentServer {
    private final Project project;
    private final ObjectMapper mapper = new ObjectMapper();
    private final InMemoryPendingApprovalService pendingApprovalService = new InMemoryPendingApprovalService();
    private final LiteToolExecutor toolExecutor;
    private final LiteAgentEngine engine;
    private HttpServer server;
    private int port;

    public LiteAgentServer(Project project) {
        this.project = project;
        this.toolExecutor = new LiteToolExecutor(project, mapper, pendingApprovalService);
        this.engine = new LiteAgentEngine(project, mapper, pendingApprovalService, toolExecutor);
    }

    public void start() {
        try {
            HttpServer httpServer = HttpServer.create(new InetSocketAddress(InetAddress.getLoopbackAddress(), 0), 32);
            httpServer.createContext("/api/agent/chat/stream", this::handleStream);
            httpServer.createContext("/api/agent/skip", this::handleSkip);
            httpServer.createContext("/api/agent/confirm", this::handleConfirm);
            httpServer.createContext("/api/agent/pending-command", this::handlePendingCommand);
            httpServer.createContext("/api/agent/todos", this::handleTodos);
            httpServer.setExecutor(Executors.newCachedThreadPool(r -> {
                Thread thread = new Thread(r, "rikki-lite-agent");
                thread.setDaemon(true);
                return thread;
            }));
            httpServer.start();
            server = httpServer;
            port = httpServer.getAddress().getPort();
            System.setProperty("rikki.endpoint", "http://127.0.0.1:" + port + "/api/agent/chat/stream");
            System.setProperty("rikki.skip.endpoint", "http://127.0.0.1:" + port + "/api/agent/skip");
            System.setProperty("rikki.confirm.endpoint", "http://127.0.0.1:" + port + "/api/agent/confirm");
            System.setProperty("rikki.pending.command.endpoint", "http://127.0.0.1:" + port + "/api/agent/pending-command");
            System.setProperty("rikki.pending.enabled", "false");
        } catch (Exception e) {
            throw new RuntimeException("Failed to start LiteAgentServer", e);
        }
    }

    public void stop() {
        if (server != null) {
            server.stop(0);
            server = null;
        }
    }

    public int getPort() {
        return port;
    }

    private void handleSkip(HttpExchange exchange) throws java.io.IOException {
        pendingApprovalService.skipCurrentExecution();
        exchange.sendResponseHeaders(200, -1);
        exchange.close();
    }

    private void handleConfirm(HttpExchange exchange) throws java.io.IOException {
        String query = exchange.getRequestURI().getRawQuery();
        boolean approved = query != null && query.toLowerCase().contains("decision=approve");
        pendingApprovalService.resolveLatestPending(approved);
        exchange.sendResponseHeaders(200, -1);
        exchange.close();
    }

    private void handlePendingCommand(HttpExchange exchange) throws java.io.IOException {
        if (!"POST".equalsIgnoreCase(exchange.getRequestMethod())) {
            exchange.sendResponseHeaders(405, -1);
            exchange.close();
            return;
        }
        JsonNode request;
        try {
            request = mapper.readTree(exchange.getRequestBody().readAllBytes());
        } catch (Exception e) {
            writeJson(exchange, 400, Map.of("status", "error", "error", "Invalid JSON payload"));
            return;
        }
        Object result = pendingApprovalService.resolve(
                request.path("commandId").asText(""),
                request.path("reject").asBoolean(false),
                request.path("decisionMode").asText("")
        );
        writeJson(exchange, 200, result);
    }

    private void handleTodos(HttpExchange exchange) throws java.io.IOException {
        if (!"GET".equalsIgnoreCase(exchange.getRequestMethod())) {
            exchange.sendResponseHeaders(405, -1);
            exchange.close();
            return;
        }
        String workspaceRoot = queryParams(exchange.getRequestURI().getRawQuery()).getOrDefault("workspaceRoot", "");
        String json = workspaceRoot.isBlank() ? "[]" : toolExecutor.todosAsListJson(workspaceRoot);
        byte[] bytes = json.getBytes(StandardCharsets.UTF_8);
        exchange.getResponseHeaders().add("Content-Type", "application/json; charset=UTF-8");
        exchange.sendResponseHeaders(200, bytes.length);
        exchange.getResponseBody().write(bytes);
        exchange.getResponseBody().close();
    }

    private void handleStream(HttpExchange exchange) throws java.io.IOException {
        if (!"POST".equalsIgnoreCase(exchange.getRequestMethod())) {
            exchange.sendResponseHeaders(405, -1);
            exchange.close();
            return;
        }
        JsonNode request;
        try {
            request = mapper.readTree(exchange.getRequestBody().readAllBytes());
        } catch (Exception ignored) {
            exchange.close();
            return;
        }
        String goal = request.path("goal").asText("");
        String workspaceRoot = request.path("workspaceRoot").asText(project.getBasePath() == null ? "" : project.getBasePath());
        JsonNode ideContext = request.path("ideContext");
        JsonNode history = request.path("history");
        String sessionId = request.path("sessionID").asText("");
        JsonNode settings = request.path("settings");

        exchange.getResponseHeaders().add("Content-Type", "text/event-stream; charset=UTF-8");
        exchange.getResponseHeaders().add("Cache-Control", "no-cache");
        exchange.getResponseHeaders().add("Connection", "keep-alive");
        exchange.sendResponseHeaders(200, 0);
        LiteSseWriter writer = new LiteSseWriter(exchange.getResponseBody());
        try {
            engine.run(goal, workspaceRoot, ideContext, history, settings, sessionId, writer);
        } catch (Exception ignored) {
        } finally {
            exchange.getResponseBody().close();
        }
    }

    private void writeJson(HttpExchange exchange, int status, Object payload) throws java.io.IOException {
        byte[] bytes = mapper.writeValueAsBytes(payload);
        exchange.getResponseHeaders().add("Content-Type", "application/json; charset=UTF-8");
        exchange.sendResponseHeaders(status, bytes.length);
        exchange.getResponseBody().write(bytes);
        exchange.getResponseBody().close();
    }

    private Map<String, String> queryParams(String rawQuery) {
        HashMap<String, String> params = new HashMap<>();
        if (rawQuery == null || rawQuery.isBlank()) {
            return params;
        }
        for (String pair : rawQuery.split("&")) {
            int idx = pair.indexOf('=');
            if (idx < 0) {
                continue;
            }
            params.put(pair.substring(0, idx), URLDecoder.decode(pair.substring(idx + 1), StandardCharsets.UTF_8));
        }
        return params;
    }
}

File: idea-plugin/src/main/kotlin/com/zzf/rikki/idea/agent/LiteAgentServer.kt
package com.zzf.rikki.idea.agent

import com.fasterxml.jackson.databind.ObjectMapper
import com.intellij.openapi.project.Project
import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import com.zzf.rikki.idea.agent.compat.InMemoryPendingApprovalService
import com.zzf.rikki.idea.agent.compat.LiteToolExecutor
import kotlinx.coroutines.runBlocking
import java.net.InetAddress
import java.net.InetSocketAddress
import java.nio.charset.StandardCharsets
import java.util.concurrent.Executors

/**
 * Embedded JDK HTTP server that serves the same SSE API as the Spring Boot backend.
 * ChatPanel connects to this instead of localhost:18080.
 */
class LiteAgentServer(private val project: Project) {
    private val mapper = ObjectMapper()
    private val pendingApprovalService = InMemoryPendingApprovalService()
    private val toolExecutor = LiteToolExecutor(project, mapper, pendingApprovalService)
    private var server: HttpServer? = null
    var port: Int = 0
        private set

    fun start() {
        val srv = HttpServer.create(InetSocketAddress(InetAddress.getLoopbackAddress(), 0), 32)
        srv.createContext("/api/agent/chat/stream", ::handleStream)
        srv.createContext("/api/agent/skip", ::handleSkip)
        srv.createContext("/api/agent/confirm", ::handleConfirm)
        srv.createContext("/api/agent/pending-command", ::handlePendingCommand)
        srv.createContext("/api/agent/todos", ::handleTodos)
        srv.executor = Executors.newCachedThreadPool { r ->
            Thread(r, "rikki-lite-agent").also { it.isDaemon = true }
        }
        srv.start()
        server = srv
        port = srv.address.port
        System.setProperty("rikki.endpoint", "http://127.0.0.1:$port/api/agent/chat/stream")
        System.setProperty("rikki.skip.endpoint", "http://127.0.0.1:$port/api/agent/skip")
        System.setProperty("rikki.confirm.endpoint", "http://127.0.0.1:$port/api/agent/confirm")
        System.setProperty("rikki.pending.command.endpoint", "http://127.0.0.1:$port/api/agent/pending-command")
        System.setProperty("rikki.pending.enabled", "false")
    }

    fun stop() {
        server?.stop(0)
        server = null
    }

    private fun handleSkip(exchange: HttpExchange) {
        pendingApprovalService.skipCurrentExecution()
        exchange.sendResponseHeaders(200, -1)
        exchange.close()
    }

    private fun handleConfirm(exchange: HttpExchange) {
        val query = exchange.requestURI.rawQuery ?: ""
        val approved = query.split("&").any { it.equals("decision=approve", ignoreCase = true) }
        pendingApprovalService.resolveLatestPending(approved)
        exchange.sendResponseHeaders(200, -1)
        exchange.close()
    }

    private fun handlePendingCommand(exchange: HttpExchange) {
        if (!exchange.requestMethod.equals("POST", ignoreCase = true)) {
            exchange.sendResponseHeaders(405, -1)
            exchange.close()
            return
        }
        val request = try {
            mapper.readTree(exchange.requestBody.readAllBytes())
        } catch (_: Exception) {
            writeJson(exchange, 400, mapOf("status" to "error", "error" to "Invalid JSON payload"))
            return
        }
        val commandId = request.path("commandId").asText("")
        val reject = request.path("reject").asBoolean(false)
        val decisionMode = request.path("decisionMode").asText("")
        val result = pendingApprovalService.resolve(commandId, reject, decisionMode)
        writeJson(exchange, 200, result)
    }

    private fun handleTodos(exchange: HttpExchange) {
        if (!exchange.requestMethod.equals("GET", ignoreCase = true)) {
            exchange.sendResponseHeaders(405, -1)
            exchange.close()
            return
        }
        val workspaceRoot = queryParams(exchange.requestURI.rawQuery)["workspaceRoot"].orEmpty()
        val json = if (workspaceRoot.isBlank()) {
            "[]"
        } else {
            toolExecutor.todosAsListJson(workspaceRoot)
        }
        val bytes = json.toByteArray(StandardCharsets.UTF_8)
        exchange.responseHeaders.add("Content-Type", "application/json; charset=UTF-8")
        exchange.sendResponseHeaders(200, bytes.size.toLong())
        exchange.responseBody.use { it.write(bytes) }
    }

    private fun handleStream(exchange: HttpExchange) {
        if (!exchange.requestMethod.equals("POST", ignoreCase = true)) {
            exchange.sendResponseHeaders(405, -1)
            return
        }

        val body = try {
            exchange.requestBody.readAllBytes()
        } catch (_: Exception) {
            return
        }
        val request = try {
            mapper.readTree(body)
        } catch (_: Exception) {
            return
        }

        val goal = request.path("goal").asText("")
        val workspaceRoot = request.path("workspaceRoot").asText(project.basePath ?: "")
        val ideContext = request.path("ideContext")
        val history = request.path("history")
        val sessionId = request.path("sessionID").asText("")
        val settings = request.path("settings")

        exchange.responseHeaders.add("Content-Type", "text/event-stream; charset=UTF-8")
        exchange.responseHeaders.add("Cache-Control", "no-cache")
        exchange.responseHeaders.add("Connection", "keep-alive")
        exchange.sendResponseHeaders(200, 0)

        val writer = LiteSseWriter(exchange.responseBody)
        try {
            val engine = LiteAgentEngine(project, mapper, pendingApprovalService, toolExecutor)
            runBlocking {
                engine.run(goal, workspaceRoot, ideContext, history, settings, sessionId, writer)
            }
        } catch (_: Exception) {
        } finally {
            try {
                exchange.responseBody.close()
            } catch (_: Exception) {
            }
        }
    }

    private fun writeJson(exchange: HttpExchange, status: Int, payload: Any) {
        val bytes = mapper.writeValueAsBytes(payload)
        exchange.responseHeaders.add("Content-Type", "application/json; charset=UTF-8")
        exchange.sendResponseHeaders(status, bytes.size.toLong())
        exchange.responseBody.use { it.write(bytes) }
    }

    private fun queryParams(rawQuery: String?): Map<String, String> {
        if (rawQuery.isNullOrBlank()) {
            return emptyMap()
        }
        return rawQuery.split('&')
            .mapNotNull { pair ->
                val idx = pair.indexOf('=')
                if (idx < 0) {
                    null
                } else {
                    pair.substring(0, idx) to java.net.URLDecoder.decode(pair.substring(idx + 1), StandardCharsets.UTF_8)
                }
            }
            .toMap()
    }
}